ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,314 @@
+    #   arguments are reserved for options
+    SumActivity.execute!(a, b, start_to_close_timeout: 10)
+  end
+end
+```
+
+#### Alternative approaches
+
+- For the same exact reasons it is not feasible to use non-namespaced methods for workflow
+
+- Similar to activities it should be possible to define multiple workflows per class, however this
+  will not be allowed to avoid a perception of a shared state between workflows
+
+
+## Start a workflow
+
+```ruby
+require 'temporal/client'
+require 'sum_workflow'
+
+connection = Temporal::Connection.new('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+# workflow is referenced via a stub, not just a class
+handle = client.start_workflow(SumWorkflow.execute(1, 2), id: 'id-1', task_queue: 'demo')
+handle.result # => 3
+```
+
+#### Alternative approaches
+
+- Starting a foreign workflow from a string name
+
+```ruby
+client.start_workflow(Temporal::Workflow.execute('my.workflow', 1, 2), id: 'id-1', task_queue: 'demo')
+```
+
+- Starting a foreign workflow with the help of a stub
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  workflow 'my.workflow'
+end
+
+client.start_workflow(MyWorkflow.execute('my.workflow', 1, 2), id: 'id-1', task_queue: 'demo')
+```
+
+
+## Start a worker
+
+```ruby
+require 'temporal/worker'
+require 'sum_workflow'
+require 'sum_activity'
+require 'calculator_activities'
+
+connection = Temporal::Connection.new('localhost:7233')
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [SumWorkflow], # registers all workflows in the class
+  activities: [SumActivity, CalculatorActivities] # registers all activities in the class
+)
+
+worker.run
+```
+
+#### Alternative approach
+
+To support passing additional context to the activity classes a worker can be initialised with an
+instance of the activity class (only available for activities).
+
+```ruby
+redis = Redis.new('localhost:6379')
+
+Temporal::Worker.new(
+  ...,
+  activities: [CalculatorActivities.new(redis)]
+)
+```
+
+*NOTE: This however will come with a heavy warning — the activity must be strictly thread-safe as
+the provided instance will be called from multiple threads (and possibly at the same exact time).*
+
+
+## Activity heartbeats & cancellations
+
+```ruby
+class MyActivity < Temporal::Activity
+  def execute
+    loop do
+      sleep 1
+      activity.heartbeat('ping')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise 'this activity is cancelled'
+  end
+end
+```
+
+This approach uses `Thread#raise` that will eject the execution out of the current context. In cases
+where this is not acceptable we will allow to either wrap a critical code in an `activity.shield`
+block (that will not raise until the block has executed) or opt-out of this default behaviour
+altogether and use `activity.cancelled?` flag to determine when an activity was cancelled.
+
+Here is what it will look like from a workflow perspective:
```

Comment:
We will have to have a (maybe separate) discussion about workflow cancellation and its cascading effects. We can't use `Thread#raise` in workflows so you very well may want to turn to something like https://docs.temporal.io/typescript/cancellation-scopes. But for activity cancellation, this looks perfect.